compiler/rustc_session/src/config.rs

```diff
@@ -1061,35 +1071,114 @@ pub fn to_crate_check_config(cfg: CheckCfg) -> CrateCheckConfig {
+        for (name, values) in [
+            (sym::unix, Values::None),
+            (sym::windows, Values::None),
+            (sym::debug_assertions, Values::None),
+            (sym::proc_macro, Values::None),
+            (sym::test, Values::None),
+            (sym::doc, Values::None),
+            (sym::doctest, Values::None),
+            (sym::panic, Values::FromSlice(panic_values)),
+            (sym::feature, Values::Unknown), // This is passed by cargo
+            (sym::sanitize, Values::FromIterator(&mut sanitize_values)),
+            (sym::target_os, Values::FromIterator(&mut target_os_values)),
+            (sym::target_family, Values::FromIterator(&mut target_family_values)),
+            (sym::target_arch, Values::FromIterator(&mut target_arch_values)),
+            (sym::target_endian, Values::FromIterator(&mut target_endian_values)),
+            (sym::target_pointer_width, Values::FromIterator(&mut target_pointer_width_values)),
+            (sym::target_env, Values::FromIterator(&mut target_env_values)),
+            (sym::target_abi, Values::FromIterator(&mut target_abi_values)),
+            (sym::target_vendor, Values::FromIterator(&mut target_vendor_values)),
+            (sym::target_thread_local, Values::None),
+            (sym::target_has_atomic, Values::FromSlice(atomic_values)),
+            (sym::target_has_atomic_load_store, Values::FromSlice(atomic_values)),
+            (sym::target_has_atomic_equal_alignment, Values::FromSlice(atomic_values)),
+        ] {
+            // We only insert well-known names if `names()` was activated
+            if let Some(names_valid) = &mut self.names_valid {
                 names_valid.insert(name);
             }
+
+            // We only insert well-known values if `values()` was activated
+            if self.well_known_values {
+                match values {
+                    Values::None => {
+                        // No value is expected, so lint on every values
+                        self.values_valid.entry(name).or_insert_with(|| FxHashSet::default());
+                    }
+                    Values::Unknown => {
+                        // Because we don't know the list of values and we don't want to lint on
+                        // them, we do nothing regarding `self.values_valid`
+                    }
+                    Values::FromSlice(symbols) => {
+                        self.values_valid
+                            .entry(name)
+                            .and_modify(|values| values.extend(symbols))
+                            .or_insert_with(|| symbols.iter().map(|s| *s).collect());
+                    }
+                    Values::FromIterator(mut it) => {
+                        self.values_valid
+                            .entry(name)
+                            .and_modify(|values| values.extend(&mut it))
+                            .or_insert_with(|| it.collect());
+                    }
+                }
+            }
```

Comment:
All this logic is pretty ineffective and will bring observable slowdown when compiling hello-world like programs, given that the built-in target list may contain hundreds of targets.

I'd keep the `const WELL_KNOWN_NAMES` list as is and then do the most straightforward thing:
```rust
if self.well_known_values {
  for target in TARGETS.map(Target::builtin) {
    self.values_valid.entry(sym::target_os).or_default().insert(target.options.os);
    self.values_valid.entry(sym::targetarch).or_default().insert(target.options.arch);
    // ... etc ...
}
}
```

This way it's easier to forget to populate values for some well-known name, but if something is forgotten, then it will be noticed very quickly anyway due to warnings over the standard library and test suite (*).

(*) Is cfg-checking enabled for the test suite yet? If not, then it should be.

Reply:
Even after the optimization eager loading of all targets will bring an observable slowdown though (at least when value checking is enabled).

Reply:
> Is cfg-checking enabled for the test suite yet? If not, then it should be.

No, not yet. It's https://github.com/rust-lang/rust/pull/94298 who is currently blocked on https://github.com/rust-lang/rust/pull/94554 and me figuring out why CI is getting a warning but I don't in local.

> Even after the optimization eager loading of all targets will bring an observable slowdown though (at least when value checking is enabled).

Yes, it will. That's why I was hoping to be able to create all of that a compile-time with const but as you said it seems impossible for some target.

Reply:
You maybe interested to know that std as some use of undeclared values in the standard library:
```rust 
    (Some(Mode::Std), "target_env", Some(&["libnx"])),
    (Some(Mode::Std), "target_os", Some(&["tvos", "watchos"])),
    (
        Some(Mode::Std),
        "target_arch",
        Some(&["asmjs", "spirv", "nvptx", "nvptx64", "le32", "xtensa"]),
    ),
```
*(Taken from a earlier version of this PR and https://github.com/rust-lang/rust/pull/94298)*